README.md
# LRSwitchboard (rest-based back-end)
Code Repository for the *rest-based back-end* of the Language Resources Switchboard of CLARIN.

The implementation is based upon nodejs and the express library. The code provides a rest-based interface
to the switchboard's resource-tool matching service.

The new design separates the switchboard's front-end with those parts that find applicable tools
for a given resource. The back-end consists of

- the tool registry 
- the matcher
- and the rest api with calls to the matcher.

# Website

The switchboard is open for business at

```https://switchboard.clarin.eu ```

Now, both the production and development version of the switchboard make use of the new
design, which is served at

```https://weblicht.sfs.uni-tuebingen.de/clrs-dev/#/ ```


# Status

The rest-based interface of the LRSwitchboard is currently being developed and actively
maintained. It has no official release yet.  For more information on the switchboard, see

```https://github.com/clarin-eric/LRSwitchboard ```





Changelog.md
# Changelog
	All notable changes to this project will be documented in this file.

## [1.4.3] - 2019-16-11
	Changed
	- version now synchronised with switchboard version
	- added new languages for WebSty
	- re-added SpaCy entries (D4Science)
	- removed duplicate slots
	
## [1.1] - 2018-11-30
### Removed
	- all entries representing web services (deprecated)
	- slot softwareType reused for type of tool: quantitative/qualitative
	(although this information is not exploited yet)

## [1.0] - 2018-10-17
### Added
	- two new tools from D4Science (SpaCy German and SpaCy English)
	- new slot to mark lifecycle status of tool, if not present, production status is assumed
	(developer version of switchboard now includes tools with lifecycle 'beta')



index.js
// -------------------------------------------
// The CLARIN Language Resource Switchboard
// 2018- Claus Zinn, University of Tuebingen
// 
// File: index.js (for rest-based service)
// Time-stamp: <2018-07-26 13:52:38 (zinn)>
// -------------------------------------------


/* expose rest-based interface to the Registry/Matcher 

   several options for parameter passing:
   - query parameters: ?language=eng
   - path parameters:  :language
   - post request carrying JSON structure with parameter information

   "primary parameters": 
           mimetype of resource
           language of resource
	   authentication for tool access

   "secondary parameters", how to present content 
           order by task
	   order by toolName

 */

const express = require('express')
const app = express()
const Matcher = require("./Matcher.js").default;

// Constants
const PORT = 8080;
const HOST = '0.0.0.0';

app.get('/doc/exampleJSON_page1.png', function (req, res) {
    res.sendfile('./exampleJSON_page1.png');    
})

app.get('/doc/exampleJSON_page2.png', function (req, res) {
    res.sendfile('./exampleJSON_page2.png');    
})

app.get('/doc/exampleResult.json', function (req, res) {
    res.sendfile('./exampleResult.json');    
})

app.get('/', function (req, res) {
    res.sendfile('./index.html');
})
    
app.get('/api', function (req, res) {
    res.sendfile('./index.html');    
})

// localhost:4511/api/tools?language=eng&includeWS=true&mimetype=application%2Fpdf&sortBy=tasks&includeToolsReqAuth=true

/* parameters:

   language:   ISO 639-3
   mimetype:   following mimetype classification given in [REF]
   includeWS:  include web services: yes or no (default=yes)
   includeToolsReqAuto: include tools (or web services) that require authentication: yes or no (default=yes)
   sortBy:     sort results by tasks (task-oriented view) or tools (tool-oriented view). (defaults=tasks)

 */

app.get('/api/tools/', function(req, res) {
    console.log('/api/tools', req);    
    res.send( new Matcher().findApplicableTools( req.query ) );    
});

/*

Tool ids are assigned during matching. They are not persistent. Might want to give each tool
a hardwired id when registered with the registry (todo)

app.get('/api/tool/', function(req, res) {
    console.log('/api/tools', req);    
    res.send( new Matcher().findApplicableTools( req.query ) );    
});
*/

app.listen(PORT, HOST);

console.log(`Running on http://${HOST}:${PORT}`);


/* Example for path parameters (not supported)

app.get('/api/tools/:language-:mediatype',function(req,res){
    const resourceDescription =
	{
	    language: {
		threeLetterCode: req.params.language },
	    mimetype: req.params.mediatype
	};
    console.log('tools', resourceDescription);
    const matcher = new Matcher();
    const applicableTools = matcher.findApplicableTools( resourceDescription, false);
    console.log('tools', applicableTools);
    res.send( applicableTools )
});

*/

Matcher.js
// -------------------------------------------
// The CLARIN Language Resource Switchboard
// 2016-18 Claus Zinn, University of Tuebingen
// 
// File: Matcher.js (for rest-based service)
// Time-stamp: <2018-10-16 15:26:12 (zinn)>
// -------------------------------------------

import uuid from 'node-uuid';
import Registry from './Registry.js';

export default class Matcher {

    constructor() {
	this.registeredTools = Registry;
//	this.groupTools = this.groupTools.bind(this);
    }

    // construct a dictionary to group all tools in terms of the tasks they can perform
    // key: task, value: tools
    groupTools( tools ){

	var toolGroups = {};

	for (var i = 0; i<tools.length; i++) {
	    const entry = tools[i];

	    // defines the registry entries visible to the outside
	    const toolInfo = [ {
		name            : entry.name,
		logo            : entry.logo,
		description     : entry.description,
		homepage        : entry.homepage,
		url             : entry.url,
		location        : entry.location,
		authentication  : entry.authentication,
		id              : entry.id,
		email           : entry.contact.email,
		parameters      : entry.parameters,
		langEncoding    : entry.langEncoding,
		output          : entry.output,
		softwareType    : entry.softwareType,
		lifecycle       : entry.lifecycle,		
		requestType     : entry.requestType,
		mapping         : entry.mapping,
	    } ];

	    if (entry.task in toolGroups) {
		toolGroups[ entry.task ] = toolGroups[ entry.task ].concat( toolInfo );
	    } else {
		toolGroups[ entry.task ] = [].concat( toolInfo );		
	    }
	}
	return toolGroups;
    }

    // todo, see comments in index.js

    /*
    getTool( query ) {
	var selectedTools = this.registeredTools;
	for (var key in query){
	    
	    console.log('key/val', key, query[key]);
	    if ( key == "id" ) {
		selectedTools = selectedTools.filter(
		    (tool) =>  {
			if (tool.id.indexOf(query[key]) != -1) {
			    return tool;
			}
		    });
	    }
	}
	return selectedTools;
    }
    */
	    
    findApplicableTools( query ) {
	
	var selectedTools = this.registeredTools;
	var sortToolsP    = true;

	console.log('----- new query', query);	
	for (var key in query){
	    
	    console.log('key/val', key, query[key]);

	    if ( key == "mimetype" ) {
		selectedTools = selectedTools.filter(
		    (tool) =>  {
			if (tool.mimetypes.indexOf(query[key]) != -1) {
			    tool.id = uuid.v4();			
			    return tool;
			}
		    });
	    }
	    
	    if ( key == "language" ) {
		selectedTools = selectedTools.filter(
		    (tool) =>   {
			if ( (tool.languages.indexOf(query[key]) != -1) ||
			     (tool.languages.indexOf("generic")  != -1)) {
			    tool.id = uuid.v4();
			    console.log('found tool', tool.name);
			    return tool;
			} 
		    });
	    }
	    
	    if (key == "includeWS" ) {
		if (query[key] == "no" ) {
		    selectedTools = selectedTools.filter(
			(tool) =>   {
			    if ((tool.softwareType == undefined) ||
				(! (tool.softwareType == "webService")) ) {
				tool.id = uuid.v4();			    
				return tool;
			    }
			});		
		}
	    }

	    if (key == "includeBetaSoftware" ) {
		if (query[key] == "no" ) {
		    selectedTools = selectedTools.filter(
			(tool) =>   {
			    if ((tool.lifecycle == undefined) ||
				(! (tool.lifecycle == "beta")) ) {
				tool.id = uuid.v4();			    
				return tool;
			    }
			});		
		}
	    }

	    if (key == "includeToolsReqAuth" ) {
		if (query[key] == "no" ) {
		    selectedTools = selectedTools.filter(
			(tool) =>   {
			    if (! (tool.authentication == "yes")) {
				tool.id = uuid.v4();			    
				return tool;
			    }
			});		
		}
	    }	    

	    // default is tasks
	    if (key == "sortBy" ) {
		if (query[key] == "tools" ) {
		    sortToolsP = false;
		}
	    }
	}

	// after all filtering operations, do the grouping
	if (sortToolsP) {
	    selectedTools = this.groupTools(selectedTools);
	}

	return selectedTools;
    }
}
